zokrates_bellman/src/groth16.rs

```diff
@@ -92,80 +93,107 @@ impl<T: Field + BellmanFieldExtensions> NonUniversalBackend<T, G16> for Bellman
-        delta: parse_g2::<T>(&parameters.vk.delta_g2),
-        gamma_abc: parameters
-            .vk
-            .ic
-            .iter()
-            .map(|g1| parse_g1::<T>(g1))
-            .collect(),
+impl<T: Field + BellmanFieldExtensions> MpcBackend<T, G16> for Bellman {
+    fn initialize<R: Read, W: Write, I: IntoIterator<Item = Statement<T>>>(
+        program: ProgIterator<T, I>,
+        phase1_radix: &mut R,
+        output: &mut W,
+    ) -> Result<(), String> {
+        let circuit = Computation::without_witness(program);
+        let params = MPCParameters::new(circuit, phase1_radix).map_err(|e| e.to_string())?;
+        params.write(output).map_err(|e| e.to_string())?;
+        Ok(())
+    }
+
+    fn contribute<R: Read, W: Write, G: Rng>(
+        params: &mut R,
+        rng: &mut G,
+        output: &mut W,
+    ) -> Result<[u8; 64], String> {
+        let mut params =
+            MPCParameters::<T::BellmanEngine>::read(params, true).map_err(|e| e.to_string())?;
+
+        let hash = params.contribute(rng);
+        params.write(output).map_err(|e| e.to_string())?;
+
+        Ok(hash)
+    }
+
+    fn verify<P: Read, R: Read, I: IntoIterator<Item = Statement<T>>>(
+        params: &mut P,
+        program: ProgIterator<T, I>,
+        phase1_radix: &mut R,
+    ) -> Result<Vec<[u8; 64]>, String> {
+        let params =
+            MPCParameters::<T::BellmanEngine>::read(params, true).map_err(|e| e.to_string())?;
+
+        let circuit = Computation::without_witness(program);
+        let hashes = params
+            .verify(circuit, phase1_radix)
+            .map_err(|_| "parameters malformed".to_string())?;
+
+        Ok(hashes)
+    }
+
+    fn export_keypair<R: Read>(params: &mut R) -> Result<SetupKeypair<T, G16>, String> {
+        let params =
+            MPCParameters::<T::BellmanEngine>::read(params, true).map_err(|e| e.to_string())?;
+
+        let params = params.get_params();
+        let mut pk: Vec<u8> = Vec::new();
+        params.write(&mut pk).map_err(|e| e.to_string())?;
+
+        let vk = serialization::parameters_to_verification_key::<T>(params);
+        Ok(SetupKeypair::new(vk, pk))
+    }
+}
+
+pub mod serialization {
+    use super::*;
+    use pairing::from_hex;
+    use zokrates_proof_systems::{G1Affine, G2Affine};
+
+    pub fn parameters_to_verification_key<T: Field + BellmanFieldExtensions>(
+        parameters: &Parameters<T::BellmanEngine>,
+    ) -> VerificationKey<G1Affine, G2Affine> {
+        VerificationKey {
+            alpha: parse_g1::<T>(&parameters.vk.alpha_g1),
+            beta: parse_g2::<T>(&parameters.vk.beta_g2),
+            gamma: parse_g2::<T>(&parameters.vk.gamma_g2),
+            delta: parse_g2::<T>(&parameters.vk.delta_g2),
+            gamma_abc: parameters
+                .vk
+                .ic
+                .iter()
+                .map(|g1| parse_g1::<T>(g1))
+                .collect(),
+        }
+    }
+
+    pub fn to_g1<T: BellmanFieldExtensions>(
```

Comment:
I think we should have `to_g1` and `to_g2` in `lib.rs` because they do not depend on the scheme.